### PR TITLE
Use of bang notation inspection

### LINKED ISSUE #4406: ImplicitDefaultMemberAssignmentInspection should ignore bang notation.
While "technically" correct, bang notation is a shorthand *intended to* access the default member. Use of bang notation in general is smelly and should be inspected IMHO, but it needs its own inspection so it can be configured independently.

### PR DESCRIPTION
**Pull request by MDoerner**
Closes #4406 
Deals with the non-control part of issue #5093 

This PR introduces the `UseOfBangNotationInspection`, which finds all uses of the bang operator. It also provides a quick fix that turns the bang notation into an explicit default member access, passing the argument as a string.

### CODE REVIEW

File: Rubberduck.CodeAnalysis/Inspections/Abstract/IdentifierReferenceInspectionBase.cs [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+        protected readonly IDeclarationFinderProvider DeclarationFinderProvider;
+
+        public IdentifierReferenceInspectionBase(RubberduckParserState state)
+            : base(state)
+        {
+            DeclarationFinderProvider = state;
+        }
+
+        protected abstract bool IsResultReference(IdentifierReference reference);
+        protected abstract string ResultDescription(IdentifierReference reference);
```
Comment (retailcoder) [inline Rubberduck.CodeAnalysis/Inspections/Abstract/IdentifierReferenceInspectionBase.cs]: You're onto something here... love it!!

File: Rubberduck.CodeAnalysis/Inspections/Concrete/UseOfRecursiveBangNotationInspection.cs [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    /// <![CDATA[
+    /// Public Function MyName(ByVal rst As ADODB.Recordset) As Variant
+    ///     MyName = rst("Name").Value
+    /// End Function
+    /// ]]>
+    /// </example>
+    /// <example hasResult="false">
+    /// <![CDATA[
+    /// Public Function MyName(ByVal rst As ADODB.Recordset) As Variant
+    ///     MyName = rst.Fields!Name.Value
```
Comment (retailcoder) [inline Rubberduck.CodeAnalysis/Inspections/Concrete/UseOfRecursiveBangNotationInspection.cs]: Is it legal?

File: Rubberduck.Resources/Inspections/InspectionInfo.resx [hunk 3]
```diff
@@ -403,4 +403,13 @@ If the parameter can be null, ignore this inspection result; passing a null valu
   <data name="DefaultMemberRequiredInspection" xml:space="preserve">
     <value>The VBA compiler does not raise an error if an indexed default member call is required but the object's declared type does not have a suitable default member. Under almost all circumstances, this leads to a run-time error 91 'Object or With block variable not set' or 438 'Object doesn't support this property or method' depending on whether the object has the value 'Nothing' or not, which is harder to detect and indicates a bug.</value>
   </data>
+  <data name="UseOfBangNotationInspection" xml:space="preserve">
+    <value>Bang notation, formally known as dictionary access expression, looks like it is strongly typed. However, it is actually a stringly  typed access to the paramterized default member of the object it is used on.</value>
```
Comment (retailcoder) [inline Rubberduck.Resources/Inspections/InspectionInfo.resx]:
```suggestion
    <value>Bang notation, formally known as dictionary access expression, looks like it is strongly typed. However, it is actually a stringly-typed access to the parameterized default member of the object it is used on.</value>
```

File: Rubberduck.Resources/Inspections/InspectionInfo.resx [hunk 4]
```diff
@@ -403,4 +403,13 @@ If the parameter can be null, ignore this inspection result; passing a null valu
...
+  </data>
+  <data name="UseOfRecursiveBangNotationInspection" xml:space="preserve">
+    <value>Bang notation, formally known as dictionary access expression, looks like it is strongly typed. However, it is actually a stringly  typed access to the paramterized default member of the object it is used on. This is especially misleading if the parameterized default member is not on the object itself and can only be reached by calling the parameterless default member first. </value>
```
Comment (retailcoder) [inline Rubberduck.Resources/Inspections/InspectionInfo.resx]:
```suggestion
    <value>Bang notation, formally known as dictionary access expression, looks like it is strongly typed. However, it is actually a stringly-typed access to the parameterized default member of the object it is used on. This is especially misleading if the parameterized default member is not on the object itself and can only be reached by calling the parameterless default member first.</value>
```

File: Rubberduck.Resources/Inspections/InspectionInfo.resx [hunk 5]
```diff
@@ -403,4 +403,13 @@ If the parameter can be null, ignore this inspection result; passing a null valu
...
+  </data>
+  <data name="UseOfUnboundBangNotationInspection" xml:space="preserve">
+    <value>Bang notation, formally known as dictionary access expression, looks like it is strongly typed. However, it is actually a stringly  typed access to the paramterized default member of the object it is used on. This is especially misleading the default member cannot be determined at compile time.</value>
```
Comment (retailcoder) [inline Rubberduck.Resources/Inspections/InspectionInfo.resx]:
```suggestion
    <value>Bang notation, formally known as dictionary access expression, looks like it is strongly typed. However, it is actually a stringly-typed access to the parameterized default member of the object it is used on. This is especially misleading the default member cannot be determined at compile-time.</value>
```

File: Rubberduck.Resources/Inspections/InspectionResults.resx [hunk 6]
```diff
@@ -458,4 +458,16 @@ In memoriam, 1972-2018</value>
     <value>The expression '{0}' requires a default member access, but the type '{1}' does not have a suitable default member.</value>
     <comment>{0} expression; {1} type</comment>
   </data>
+  <data name="UseOfBangNotationInspection" xml:space="preserve">
+    <value>The expression '{0}' contains a use of bang notation.</value>
```
Comment (retailcoder) [inline Rubberduck.Resources/Inspections/InspectionResults.resx]:
```suggestion
    <value>The expression '{0}' uses bang notation.</value>
```

File: Rubberduck.Resources/Inspections/InspectionResults.resx [hunk 7]
```diff
@@ -458,4 +458,16 @@ In memoriam, 1972-2018</value>
...
+    <comment>{0} expression</comment>
+  </data>
+  <data name="UseOfRecursiveBangNotationInspection" xml:space="preserve">
+    <value>The expression '{0}' contains a use of recursive bang notation.</value>
```
Comment (retailcoder) [inline Rubberduck.Resources/Inspections/InspectionResults.resx]:
```suggestion
    <value>The expression '{0}' uses a recursive bang notation.</value>
```

File: Rubberduck.Resources/Inspections/InspectionResults.resx [hunk 8]
```diff
@@ -458,4 +458,16 @@ In memoriam, 1972-2018</value>
...
+    <comment>{0} expression</comment>
+  </data>
+  <data name="UseOfUnboundBangNotationInspection" xml:space="preserve">
+    <value>The expression '{0}' contains a use of unbound bang notation.</value>
```
Comment (retailcoder) [inline Rubberduck.Resources/Inspections/InspectionResults.resx]:
```suggestion
    <value>The expression '{0}' uses an unbound bang operator.</value>
```

**Comment by MDoerner** [comment]
I will incorporate the suggested changes this evening old-school style on my local and push them.

**Pull request merged**